### ISSUE TITLE
Nifty asset manager

### ISSUE THREAD

**Issue by relu91** [nifty-gui/nifty-gui#411]
Hello, I saw that in the last weekly meting @mkaring mention about this feature for Nifty 2.0 and since I was also thinking about this in the past weeks, I write down my thoughts in this issue. I know it's a bit early but I promise to leave the future generation of nifty developers!!! :smile: 

Looking in the existing game frameworks/engine the interface offered by a entity that has the role of Asset Manager has simple one method to explicit load an asset one to retrive and an other to unload it.  Someone offers also a way to asynchronous functionality to avoid block in the render process. They also has some flexibility to load custom data using specific classes that parses data from files. 

What I suggest is to delegate all loading process to an interface called NiftyAssetManager that has at least two methods to retrive NiftyImage and NiftyFont but also offers an other method to retrive a general custom asset. The reason to add this particular method is that now with Nodes an user could implement a node to Render for example some special asset ( i.e. .svg files ? ) and providing a unified interface could be handy. 

Furthermore use the mechanism offered by the rendering platform could lead in a performance boost since nifty will share TextureAtlas or other resources. Finally it will lead to less dependencies in nifty-core and a more slim Nifty class interface ( remove loadNiftyImage and loadNiftyFont methods).

I volunteer to provide a first draft in a pull request.

**Comment by void256** [comment]
What's wrong with the [NiftyResourceLoader](https://github.com/nifty-gui/nifty-gui/blob/2.0/nifty/src/main/java/de/lessvoid/niftyinternal/NiftyResourceLoader.java) and it's plugable [NiftyResourceLocation](https://github.com/nifty-gui/nifty-gui/blob/2.0/nifty/src/main/java/de/lessvoid/niftyinternal/NiftyResourceLocation.java)?

Besides that the latter is not yet part of the public API :wink:

The way I see it all we need to do is to make NiftyResourceLocation part of the SPI and allow users to provide it like the other SPI impls when creating a Nifty instance. Probably making the NiftyResourceLocation an optional parameter of some sort.

**Comment by relu91** [comment]
Here a use case that it think is useful to understand the limit of current implementation of asset loading. 
Assuming we have a file called asset.xy that is shared between nifty instance and my game engine. Game engine exposes a class able to cache assets loaded and get them using a method like Object getAsset(string name). If I understand NiftyResourceLoader I would add a  NiftyResourceLocation that can call getAsset method. NiftyResourceLocation can only returns a InputStream or a Url so I need to implement a method to traslate my object instance loaded in memory again in InputStream. Assuming that the asset.xy is a image we are reading twice the texture data one in game engine and one to loading back in Nifty while we could simply give the loaded instance whithout read again a stream. 

I think this is at least a little tricky and we could provide a better and clean interface to load asset. Thus we can share information from nifty and game engine. Also with a good interface if we load an asset from nifty it's also available back to game engine side, right now it's not possibile ( at least in a clean way). Finally with the current NiftyResourceLoader there's no reference count of loaded resources, in a way that a component could say that wants to unload an asset and the asset maganer only unload it after the reference count is 0. Using the underlining asset systems can make this feature for free. 

So I would prefer to refactor a little bit the process, I don't know if I'm only a crazy man :smile:

**Comment by relu91** [comment]
Of course if there's no asset management support in the target system (i.e LWJGL) implement a bake-end for that platform it's a little bit more demanding, but we could easily provide a default asset management system :+1:

**Comment by mkaring** [comment]
I really agree with @relu91 here. Giving Nifty the ability to hook into any assent management that is provided by a backing engine is really required in my opinion. There is no way to predict how a more complex engine handles it's assets. Could be from resource streams as much as from any strange custom container files that do ... things. Also some backing engines come with dedicated support to load different types of images and such, possibility more types then what is supported by Nifty. (strange *.dds files anyone?). Proper SPI interfaces that allow loading the data Nifty requires in a custom way, will resolve such issues. 

I also don't see a problem with providing a simple implementation to this SPI as part of the core library of Nifty, since it is unlikely to be very large or a drag on the performance (when unused). So renderers for libraries without asset management can use the default implementation and renderers for libraries with asset managements can use a custom implementation.

**Comment by bgroenks96** [comment]
I agree with @relu91 and @mkaring here. Nifty should be able to share already loaded/processed data with any engine it's being used with.

**Comment by relu91** [comment]
Just leaving a comment to let you know that I'm working with it, but I'm not yet very satisfied of the result. I hope to came with a very clean solution that allows user to plug in the asset system very easily and also take advantage of the already existing implementation across the frameworks...

**Comment by mkaring** [comment]
@relu91 Do you have your wip already commited anywhere? So we can have a look and maybe give some early input?

**Comment by relu91** [comment]
Well the quest for Asset Management in nifty-gui it's a little harder than I thought :smile: . I'm starting again from scratch... . It's quite easy to find a good abstraction layer for other asset systems, but I can't find a way to integrate it in nifty. I'll write here some thoughts that I have randomly this days about this problem and I would like to hear what you guys think about them. 

Right now in nifty we have NiftyImage and NiftyFont. I expect that this list of classes that represent a kind of media will grow in the future, isn't it? (i.e. NiftySound, NiftyMusic, Nifty3DModel etc.. ). What I'm asking to myself is: **Which could the role of a Asset System be**?  Thinking about NiftyImage, it needs a NiftyTexture to loaded and also NiftyFont need some of low level resources to be loaded. So the Asset Management system could be the one who knows how to load and share this low level resource between the engine and (let me say) NiftyMedia classes. Do you agree? 

I don't yet have a look to the new DevianArt example made by void, but I think that it could be a nice training for this kind of system, where all the loading logic is delegated to this of component. What I'm afraid of is that the current NiftyImage and NiftyFont should be changed a bit to fit in this kind of system. Also I should think about an extensible system where other NiftyMedia classes could be added and easly loaded, created and then disposed. 
The **goal** is basically to have a store for engine asset resources and than an inteface where NiftyMedia clasess could be created using them in a transparent way. Do you like this idea? 

I'm really open for discussions ! :+1:

**Comment by void256** [comment]
The DeviantArt example is still a bit rough :smile: but it can act as a real life example for all sorts of things. And we should add more of that kind of examples in the future too.

I can't really comment that much on an Asset System since I still don't know enough about them but here are a few thoughts that might eventually help in the discussion:
- NiftyImage and NiftyFont are not really the right abstraction level to be viewed as an asset in my opinion. I view them more as the public API or a handle to the underlying actual resources. So even though they appear to you as the actual resource they are merely a wrapper (in a way) for them. Of course at some point someone needs to instantiate a NiftyImage for instance and to do that the actual resource is needed and I'd say that at that point some Asset Managment System needs to be called.
- For NiftyImage the actual resource is the NiftyTexture of the SPI. These are implementation specific resources to whatever RenderDevice implementation one uses.
- NiftyFont is currently not using any SPI at all. Currently it's backing resource is a JGLFont since the jglfont project is used. However the JGLFontRenderer (the SPI of the jgfont project to actual render a JGLFont) is implemented back in Nifty using NiftyRenderDevice. The resource data for fonts is provided by the jglfont project (since all the bitmap font parsing is happening there and so on). I'm not sure yet where something like an Asset Manager would fit in but when we keep using JGLFont then we need to find a way to introduce some Asset Manager there (since again, the actual resources are accessed somewhere in the jglfont project)

So I would see an AssetManager mechanism more like close to the managment of the actual underlying resources (NiftyTexture in the NiftyImage case) and not so much as a way to replace NiftyImage.

I think we just need a way to somewhat transparently SPI like identify and provide resources to Nifty. You would still talk to the same NiftyImage API but internally the actual data can come from whatever asset store thing you'd like to use.

Disclaimer: This is not a well thought out solution to the problem but merely some random thoughts about it :wink:
